### PR TITLE
plugins: linux-vaapi, Linux VA-API H.264 encoder

### PR DESCRIPTION
**Pull request by kingargyle**
Includes code from the reboot/obs-studio github project and the vaapi-h264 branch.

This adds the ability to use libva to use the gpu to do hardware encoding.  It is 
functional for both streaming and file recording.

Tested with Intel Haswell chipset.

Note that I had contact @reboot in regards to this code as it came from his branch.  According to him it was from another branch in the original obs-studio code, but that branch has since been deleted and the work was never included into the original code.   It also appears that @kc5nra did a lot of the original work.

I'm re-submitting it for possible inclusion if somebody wants to take it over.   I believe it addresses issue 
https://obsproject.com/mantis/view.php?id=263

### CODE REVIEW

**Comment by kingargyle** [comment]
Unfortunately work is blocking the travis-ci build link, I'll have to check to see why that failed this weekend.

**Comment by kc5nra** [comment]
Hey David,

Let me know if if you have any questions about this code!  I think the only
remaining was adding b-day support (more complicated logic for setting
which frames ref each other in the gop, since now it is pretty easy and
just increments with a single ref to ref frame)

**Comment by kingargyle** [comment]
@kc5nra It is pretty functional as it is right now from my limited testing.  I mainly wanted to get this back out to a wider audience, as it really made a huge difference on my dedicated streaming laptop I take to tournaments with me.   I might not have time myself to work on it but figure the wider community could help out to finish it up.

**Comment by kingargyle** [comment]
This looks like it is missing the libva-dev files installed on the travis build system.

In file included from /home/travis/build/jp9000/obs-studio/plugins/linux-vaapi/surface-queue.c:1:0:
/home/travis/build/jp9000/obs-studio/plugins/linux-vaapi/surface-queue.h:3:19: fatal error: va/va.h: No such file or directory
 #include <va/va.h>
                   ^
compilation terminated.

It can't find the appropriate header files on the system.

File: plugins/linux-vaapi/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,40 @@
+project(linux-vaapi)
+
+find_package(LibVa)
+find_package(X11 REQUIRED)
+
```
Comment (RytoEX) [inline plugins/linux-vaapi/CMakeLists.txt]: To allow graceful failure/omission of this module if libva or other dependencies are not found on the system, add this here:

```
if(NOT LIBVA_INCLUDES_FOUND AND ENABLE_VA)
	message(FATAL_ERROR "LibVA includes not found but set as enabled")
elseif(NOT LIBVA_INCLUDES_FOUND)
	message(STATUS "LibVA includes not found, VAAPI plugin disabled")
	return()
endif()
```

**Comment by kc5nra** [comment]
Haha, thanks RytoEX.  This code is _old_, though I appreciate you going through all this.

It might make more sense if I reopen this PR with RytoEX's fixes so that the origin authorship is preserved in the git history.  Let me know if this is something you are interested in!

**Comment by kingargyle** [comment]
@kc5nra yeah, I'm fine with you re-opening the PR since you originally wrote it and have the commit history.   I just didn't want this to get lost and found it useful enough to use in place of the x264 encoder.

**Comment by RytoEX** [comment]
@kc5nra No problem.  If you open a new PR for this and you want another review of that, just tag me.

**Comment by reboot** [comment]
If you reopen the PR with the original branch don't forget to include the fix i made on https://github.com/reboot/obs-studio/commits/vaapi-h264. This PR is based on my repository, as far as i know, and should contain the fix.

**Comment by kingargyle** [comment]
@reboot yes, is based on your repository.

**Comment by pschichtel** [comment]
What is currently holding this back?

**Comment by sebix** [comment]
@pschichtel One build fails on travis: https://travis-ci.org/jp9000/obs-studio/jobs/240913031#L4248 (may be a reason)

**Comment by kingargyle** [comment]
Actually, I think @kc5nra was going to re-submit the PR since he originally worked on it, and it would contain the full development history.  That is the reason I haven't tried to address any of the issues above or the build failure.

**Comment by kc5nra** [comment]
Ya, debating whether to submit it with libyami which would get vp8/9,264/5
hardware encoding, or fixing mine to work with bframes.

**Comment by kingargyle** [comment]
@kc5nra I ran this code for 13+ hrs of recording this last weekend, and it worked well, expect for the occasional B-Frame glitch every once in a while.   If I had the time I'd try and investigate it myself as this worked well enough for what I needed it to do (i.e. relieve the stress on the CPU so it could do other things besides encode the video).  I'd encourage fixing the B-Frame and then working on the libyami implementation, but it is up to you.

**Comment by vsnrain** [comment]
Hello. I tried using this branch, but whenever I start recording or streaming with hardware encoder, it fails with error:
```
12:00:52 AM.413: [VAAPI encoder]: "vaEndPicture(q->display, q->context)": invalid parameter
12:00:52 AM.413: [VAAPI encoder]: unable to encode frame
12:00:52 AM.413: Error encoding with encoder 'streaming_h264'
```
Still, h264_vaapi hardware encoder seems to work ok when converting streams in ffmpeg.
What could be the problem?
Full logs [here.](https://gist.github.com/vsnrain/4478b6969def4ccf7f88525421a83493)

**Comment by kingargyle** [comment]
@vsnrain what does the command ```vainfo``` report?

**Comment by vsnrain** [comment]
@kingargyle [vainfo](https://gist.github.com/vsnrain/ce1f36886a62fcab6f0f23eb21450ee0)

**Comment by kingargyle** [comment]
@vsnrain after upgrading to 17.04 of Ubuntu, I'm seeing the same error.

**Comment by kingargyle** [comment]
More information;

info: [VAAPI encoder]: settings:

	device_type :          X11
	device:                Intel i965 driver for Intel(R) Haswell Mobile - 1.8.0
	device_path:           (null)
	fps_num:               30
	fps_den:               1
	width:                 1280
	height:                720
	profile:               main
	keyint:                2 (s)
	rc_type:               CBR
	bitrate:               2500 (kbit)
	use_custom_cpb:        no

info: libfdk_aac encoder created
info: libfdk_aac bitrate: 160, channels: 2
info: ==== Recording Start ===============================================
info: [ffmpeg muxer: 'adv_file_output'] Writing file '/home/dcarver/Videos/ravensq/2017-07-28 19-34-16.mkv'...
error: [VAAPI encoder]: "vaEndPicture(q->display, q->context)": invalid parameter
error: [VAAPI encoder]: unable to encode frame
error: Error encoding with encoder 'recording_h264'

The issue appears that the display_path is returning NULL.  I'll see if I can track this down.

**Comment by kingargyle** [comment]
Further investigation seems to be a bug either in Ubuntu 17.04 or the libva implementation there.  1.7.3-2 causes the same problem.   Ubuntu 16.10 which has 1.7.1 and corresponding intel drivers worked.  I didn't have any issues until upgrading to 17.04.

I've opened up an issue with the libva project, so maybe I can get some info from them

https://github.com/01org/libva/issues/87

**Comment by vsnrain** [comment]
I am using Arch, and I tried downgrading packages to
```
libva-intel-driver-1.7.1-1
libva-1.7.1-1
```
and it is still does not work. Same behavior, but this time it does not even show error and end log with 
```
==== Recording Start ===============================================
[ffmpeg muxer: 'adv_file_output'] Writing file '/home/vsnrain/2017-07-29 10-39-16.mkv'...
```
the video file is created, but only with single frame.

**Comment by kingargyle** [comment]
@vsnrain I'm going to try and install a dual boot into ubuntu 16.10 later tonight which has 1.7.1-2 and I'll let you know what I see.   I didn't start getting problems until I upgraded to 17.04.   I'll let people know what I find out, and hopefully can get some extra info from the libva project itself.

**Comment by kingargyle** [comment]
Locally I was able to get this working again, but I had to go back to Ubuntu 16.04 and it's native libva libraries.  This is running 1.7.0 of the intel driver.   I've tested it with FLV and MKV files.   I do think long term that lib-yami is probably a better option as it wraps VA-API and takes much of the grunt work out of it from what I can tell.   Something changed in the later versions, just not sure what yet.

**Comment by kingargyle** [comment]
So answer I got from the LIBVA developers is as follows.

"intra_period is set to 60 however ip_period is to 0 in your case, however 0 is invalid for ip_period when using CBR mode. The new driver adds some checks for intra_period/ip_period. Please set a valid ip_period, sorry for the inconvenience."

So, it looks like with the newer drivers we need to ip_period to something other than 0.

**Comment by kingargyle** [comment]
Putting this here, as a sample on on how to set the ip_period, (sample comes from the libva project).

"You may refer to https://github.com/01org/libva-utils/blob/master/encode/h264encode.c for how to set ip_period in your application:"

**Comment by Bleuzen** [comment]
I also get this error when using CBR:

```
error: [VAAPI encoder]: "vaEndPicture(q->display, q->context)": invalid parameter
error: [VAAPI encoder]: unable to encode frame
error: Error encoding with encoder 'streaming_h264'
```

So I tryed CQP, maybe I can at least use it for recording.
But no.
When recording with VAAPI CQP it freezes my computer after about 20 seconds.
I can't even press the reset button: nothing happens
I have to take it from the power cable to 'shut down' and restart it.

I made a 2 seconds test recording (because I don't want my computer to freeze again) and this is the log file of it:
[2017-08-12 09-24-29.txt](https://github.com/jp9000/obs-studio/files/1219570/2017-08-12.09-24-29.txt)

The recorded video is ok and plays finde in VLC.

**Comment by kingargyle** [comment]
yeah, this code should be considered beta.  I'm hoping to get some time soon to at least address the CBR issue that has been reported.   I've had some success with VBR but again that it is on Ubuntu 16.04.

**Comment by iopq** [comment]
I tried compiling the fork, and I can't run it. It compiles, then:

`Attempted path: ../../data/obs-studio/locale/en-US.ini
Attempted path: ../../data/obs-studio/locale.ini
Attempted path: ../../data/obs-studio/themes/Default.qss
Attempted path: ../../data/obs-studio/license/gplv2.txt
info: Processor: 4 logical cores
info: Processor: Intel(R) Core(TM) i7-4770K CPU @ 3.50GHz
info: Physical Memory: 15975MB Total
info: Kernel Version: Linux 4.10.0-32-generic
info: Distribution: "Ubuntu" "16.04"
info: Portable mode: false
info: OBS 19.0.2-21-g6d467e4 (linux)
info: ---------------------------------
info: ---------------------------------
info: audio settings reset:
	samples per sec: 44100
	speakers:        2
info: ---------------------------------
info: Initializing OpenGL...
info: OpenGL version: 4.5 (Core Profile) Mesa 17.3.0-devel - padoka PPA
info: ---------------------------------
info: video settings reset:
	base resolution:   2560x1440
	output resolution: 852x480
	downscale filter:  Bilinear
	fps:               30/1
	format:            NV12
info: ---------------------------------
libDeckLinkAPI.so: cannot open shared object file: No such file or directory
info: No blackmagic support
libva info: VA-API version 0.40.0
libva info: va_getDriverName() returns 0
libva info: Trying to open /usr/lib/x86_64-linux-gnu/dri/r600_drv_video.so
: CommandLine Error: Option 'help-list' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options`

Master compiles and runs just fine. I'd love to test this.

**Comment by iopq** [comment]
I solved the issue, it was with padoka PPA. Works fine when I purged it.

**Comment by iopq** [comment]
Correction: it started and let me select the encoder. It doesn't actually encode.

relevant part of the log:

```
07:31:05 PM.410: [VAAPI encoder]: "vaInitialize(display, &major_version, &minor_version)": unknown libva error
```

interestingly enough, it lets me try to start recording, but won't record anything. Stopping the recording usually just keep the program alive forever and you have to actually kill it.

**Comment by kingargyle** [comment]
@iopq what is the  vainfo output.  It looks like it it isn't able to even initialize your driver.

**Comment by kingargyle** [comment]
I know it works with 1.7.0 of the libva driver   newer versions may not work.

**Comment by iopq** [comment]
```
libva info: VA-API version 0.39.2
libva info: va_getDriverName() returns 0
libva info: Trying to open /usr/lib/x86_64-linux-gnu/dri/radeonsi_drv_video.so
libva info: Found init function __vaDriverInit_0_39
libva info: va_openDriver() returns 0
vainfo: VA-API version: 0.39 (libva 1.7.1)
vainfo: Driver version: mesa gallium vaapi
vainfo: Supported profile and entrypoints
      VAProfileMPEG2Simple            :	VAEntrypointVLD
      VAProfileMPEG2Main              :	VAEntrypointVLD
      VAProfileVC1Simple              :	VAEntrypointVLD
      VAProfileVC1Main                :	VAEntrypointVLD
      VAProfileVC1Advanced            :	VAEntrypointVLD
      VAProfileH264ConstrainedBaseline:	VAEntrypointVLD
      VAProfileH264ConstrainedBaseline:	VAEntrypointEncSlice
      VAProfileH264Main               :	VAEntrypointVLD
      VAProfileH264Main               :	VAEntrypointEncSlice
      VAProfileH264High               :	VAEntrypointVLD
      VAProfileH264High               :	VAEntrypointEncSlice
      VAProfileNone                   :	VAEntrypointVideoProc
```

**Comment by parataxia** [comment]
Confirming I got this to work in ArchBang Linux (basically identical to arch linux besides the fact it uses openrc and linux LTS.. same repos, however.) It works great after installing these packages:  `libvdpau-va-gl`and `lib32-libva-intel-driver ` , and of course compiling your source. Thanks to this I can stream my games unlike before. Using Intel Series 2000, I had a variable FPS of 40/60 depending on graphic processing intensity... so now I'm getting a steady 60FPS, vasts better. Thanks.

**Comment by Bleuzen** [comment]
I hadn't the `libvdpau-va-gl` installed.
So I installed it.
CBR and VBR still doesn't work and CQP still freezes my PC :/

**Comment by kingargyle** [comment]
@Bleuzen  I suspect you are hitting the bug described in this [comment](https://github.com/jp9000/obs-studio/pull/941#issuecomment-318963111) where the newer drivers are expecting a non zero value for the ip_period.  I unfortunately haven't had time to take a look at the code to make any adjustments.

**Comment by RytoEX** [comment]
@kingargyle @kc5nra @reboot
Any additional work happening here or brewing behind the scenes?

**Comment by kingargyle** [comment]
@RytoEX haven't had time myself to work on things.  I know what is causing the issue for the crashes some people are experiencing I just haven't had time to actually make the appropriate change in the code. I thought @kc5nra was going to either update this or create a different implementation.  I've got a work system locally but would like to try and fix it for others using the newer vaapi drivers.

Last implementation I did was just to sync my code up with master to make sure it still worked locally.

**Comment by jp9000** [comment]
Just making a sort of official note here -- unfortunately still not mergeable as of yet I feel.  Also still has that cmake configuration issue where it still won't find va/va.h when compiling for example.  The required package really needs to be put in cmake.

Might be worth checking out libyami like fruitman was saying.  Unfortunately I can't really dedicate the time to implement this myself -- I feel like I'll really have to delegate this to someone else who's able and willing.

I have a backup of the old branch still locally as well just in case.

**Comment by kingargyle** [comment]
@jp9000 I agree that libyami is probably the way forward.   I'll be giving a presentation in a couple weeks at the Ohio Linuxfest on OBS Studio and I'll bring up there that the project is looking for someone to step up and create an Intel hardware encoder.  Like others, my time unfortunately right now is limited and I have a version working for my needs.   If you want to close the PR, feel free to do so  I'll keep my branch up to date with master.

**Comment by jp9000** [comment]
Oh really?  That's quite awesome, wish I could be there to see it in person and say hi -- if there's a video of it email it to me.

Perfectly understandable though -- I'm very grateful regardless.  Having it as a PR put it kind of in a more public spotlight than when it was rotting on a stale branch, and the testing on the PR has been interesting.  VAAPI is certainly usable, although implementing it properly is definitely not a simple task.  A library that implements it is likely the best option at this point.

It's not really a critical or high priority thing these days as CPUs are more powerful and have more cores.  x264 doesn't really have troubles these days like it used to 4 or 5 years ago  I suppose hardware encoding is mostly just a luxury more than anything, so definitely not a big deal either way.

**Comment by pschichtel** [comment]
From what I have seen about libyami, this library seems focus on Intel's CPUs. Would this still work with other GPUs supporting vaapi ?

**Comment by kingargyle** [comment]
@jp9000 actually, I use a laptop for most of my use of OBS and honestly, I had to have hardware encoding because using the x264 encoder would eventually cause dropped frames when the CPU usage would get up around 20% on the system.  Using the VAAPI hardware encoder built into the laptop, keeps the CPU usage around 8% to 10% which doesn't cause any frame drops.    So in my case, I need the hardware encoding as x264 even on the fastest setting possible uses too much CPU for the poor laptop.

**Comment by jp9000** [comment]
I suppose that's true, I have a bad tendency to sort of disregard laptops when it comes to serious streaming -- it's kind of a bad habit.

**Comment by kingargyle** [comment]
@jp9000 No big deal, the use case I have, is recording Live events for miniature games, and the laptop was just the most portable solution.  Ideally I get the time to build out a more robust mini desktop machine that can be taken, but until then the laptop with hardware encoding does the job.

**Comment by reboot** [comment]
I have been wondering if there is any benefit of directly implementing a VAAPI encoder. FFmpeg has a h264_vaapi encoder that could probably be used just like it is already done for NVENC.

**Comment by iopq** [comment]
Any way to use the ffmpeg h265_vaapi encoder in OBS? Or is that something that would require a pull request?

**Comment by reboot** [comment]
The FFmpeg output in the advanced recording settings should allow any codec supported by FFmpeg.

**Comment by iopq** [comment]
There isn't such a setting in the streaming tab

**Comment by kingargyle** [comment]
@reboot I've tried to get the ffmpeg encoder with vaapi working in OBS and have failed miserably.  None of the command line options seem to work in the ffmpeg encoder settings.   If somebody wants to wrap a friendlier gui around it I'd be all for it, but until then, I think from a users point of view, having something native just makes it easier for those that don't want to have a PHD in FFMPeg.

**Comment by jp9000** [comment]
Well technically all you'd need to do it basically copy a significant portion of the NVENC code (which is implemented through FFmpeg similarly) and just adapt it to VAAPI, I should probably try that if I ever get time.  Could probably make that code abstract at some point like I did for the audio encoders.

**Comment by ONSBalder** [comment]
Can one of the admins verify this patch?

**Comment by kingargyle** [comment]
@ONSBalder  there are some bugs as noted with this patch that would need to be addressed so that it works with the newer VAAPI versions.  I think this is mainly staying open until either I get time to try and fix a few of the bugs, or somebody comes along and writes a better more complete implementation.

**Comment by RytoEX** [comment]
@kingargyle 
ONSBalder is a bot that had been dormat, and then suddenly awakened and commented on all of the open PRs.

**Comment by kingargyle** [comment]
Ah... thought that was the case after I checked it's activity. :)

**Comment by Bleuzen** [comment]
I took the code from w23 and updated it to work with the newest obs.
This uses ffmpeg and it works for me on Manjaro with ffmpeg 3.4 and libva 1.8.3 installed.

Old source: https://github.com/w23/obs-studio/tree/ffmpeg-vaapi
Updated branch with 20.1: https://github.com/Bleuzen/obs-studio/tree/ffmpeg-vaapi
Patch file: https://patch-diff.githubusercontent.com/raw/Bleuzen/obs-studio/pull/1.patch

The only thing that you have to change is this line in `/plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c`:
`ret = av_hwdevice_ctx_create(&enc->vadevice_ref, AV_HWDEVICE_TYPE_VAAPI, "/dev/dri/renderD129" /*FIXME*/, NULL, 0);`
for me it works with `/dev/dri/renderD129`, because `/dev/dri/renderD128` is my AMD GPU and `/dev/dri/renderD129` my Intel GPU.
If you also have 2 GPUs, you may have to change `/dev/dri/renderD129` to `/dev/dri/renderD128` before compiling.

Sadly there is currently no GUI to change the render device. I'm not a C developer, so I won't do it. But the basic code works and maybe someone can implement it into the GUI.

**Comment by eduardoc25** [comment]
Hi guys,

I've tried the @kingargyle branch and this is what throws me into the console:
`nfo: ==== Recording Start ===============================================
info: [ffmpeg muxer: 'adv_file_output'] Writing file '/home/eduardo/2017-10-26_14-56-01.mp4'...
error: [VAAPI encoder]: "vaDeriveImage(enc->display, surface, &image)": invalid VASurfaceID
error: [VAAPI encoder]: unable to upload frame to input surface
error: Error encoding with encoder 'recording_h264'`
This is the last configuration that probe, but without success in any test.
screenshot_20171026_150029
I've also tested the @Buzuzen branch and the latter works but with glitches...
`info: ==== Recording Start ===============================================
info: [ffmpeg muxer: 'adv_file_output'] Writing file '/home/eduardo/2017-10-26_15-09-08.mp4'...
[mp4 @ 0x2043280] Using AVStream.codec to pass codec parameters to muxers is deprecated, use AVStream.codecpar instead.
[mp4 @ 0x2043280] Using AVStream.codec to pass codec parameters to muxers is deprecated, use AVStream.codecpar instead.
[mp4 @ 0x2043280] track 1: codec frame size is not set
info: [ffmpeg muxer: 'adv_file_output'] Output of file '/home/eduardo/2017-10-26_15-09-08.mp4' stopped
info: Output 'adv_file_output': stopping
info: Output 'adv_file_output': Total frames output: 1775
info: Output 'adv_file_output': Total drawn frames: 1938 (1944 attempted)
info: Output 'adv_file_output': Number of lagged frames due to rendering lag/stalls: 6 (0,3%)
info: ==== Recording Stop ================================================
info: Video stopped, number of skipped frames due to encoding lag: 1665/1855 (89,8%)`
This is the last configuration that probe:
screenshot_20171026_151241
I do not know if I still have to implement B frames for AMD accelerators, since both in my R9 290 and my old 7850 I get the same message:
`libva info: VA-API version 0.40.0
libva info: va_getDriverName() returns 0
libva info: Trying to open /usr/lib64/dri/radeonsi_drv_video.so
libva info: Found init function __vaDriverInit_0_40
libva info: va_openDriver() returns 0
[h264_vaapi @ 0x2aa68c0] B frames are not supported (0x1).
warning: [VAAPI encoder: 'recording_h264'] Failed to open VAAPI codec: Argumento inválido`
This is how it is recorded with AMD cards:
[https://youtu.be/qBEZ58nl0No](url)
PS: My distribution is Opensuse Tumbleweed (desktop) and Opensuse 42.3 (notebook), I had no problem compiling both projects on my computers, I have all packages updated to date. I want to emphasize that both projects work for me as long as I use the intel accelerator (HD4600). Actually my skills as a programmer are very low so I point to test.
Regards

**Comment by eduardoc25** [comment]
Hi guys, as I do not see any new advances... I want to inform mine ...
At last I have managed to make vaapi work with my  AMD GPU through ffmpeg, these are the parameters:

`DRI_PRIME=1 LIBVA_DRIVER_NAME=radeonsi ffmpeg -hwaccel vaapi -vaapi_device /dev/dri/renderD128 -hwaccel_output_format vaapi -f x11grab -r 30 -s 1920x1080 -i :0.0+0,0 -an -f pulse -ac 2 -i 1 -vf 'format=nv12|vaapi,hwupload' -threads 4 -vcodec h264_vaapi -profile 66 -bf 0 -b 10M -acodec pcm_s16le output.mkv`

Regards

**Comment by flussence** [comment]
@eduardoc25 - If you're only interested in full screen recording, newer versions of ffmpeg have `-i kmsgrab`, which should perform much better than `x11grab` as it can skip the overhead of X11 + hwupload entirely.

**Comment by soredake** [comment]
Is it dead?

**Comment by jp9000** [comment]
We're going to use FFmpeg's implementation rather than a custom implementation.  Someone is working on it currently, just not complete yet.

**Comment by Angelspiritofswe** [comment]
Hello and here I am on Monday will I start for real


Åsa Angelspirit Eriksson

2018-02-23 21:18 GMT+01:00 soredake <notifications@github.com>:

**Comment by victimsofgaybullying** [comment]
@eduardoc25  @jp9000 
As you know all the AMD RX460-580 Polaris and up series gpu's don't support bframes. In addition to that the vaapi created steam can have problems with timestamps and packed headers.
"The Mesa [radeonsi] VAAPI implementation doesn't support packed headers"
Some upstream providers or at least right now anything using gstreamer isn't going to like the output from obs using vaapi. Someone should also bring this link to the attention of gstreamer. Also the mesa developers  should fix this before vaapi will work well on AMD or this feature is considered unstable even if implemented. 

https://bugs.freedesktop.org/show_bug.cgi?id=105277

I'm going to give obs a trial run with the vaapi patch here on intel&amdgpu.

**Comment by rkantos** [comment]
I tried building this for my x5-z8300 Intel Compute Stick using kingargyles sources and had the same issues as Bleuzen.. Going to try with his solution next..

> It's not really a critical or high priority thing these days as CPUs are more powerful and have more cores. x264 doesn't really have troubles these days like it used to 4 or 5 years ago. I suppose hardware encoding is mostly just a luxury more than anything, so definitely not a big deal either way.

Please NO! I cannot understand why people working with similar open source projects do not take the time to understand hardware encoding / decoding etc. In my opinion it is very important to understand that a lot of people are using very old hardware or in other words cheap hardware to run.. you've guessed it; open source software. This means one should forget modern hardware, such as the Intel Atoms like Bay Trail and more recently Cherry Trail CPUs which are all very capable computers for the price.

And I mean all, ALL!!, it takes is to implement proper support for at least the basic hardware capabilities of the CPUs that exist! Imagine all the Atom-based TV-devices running 2W TDP (and older hardware, like the Intel HD2000, HD2500 and HD3000) going from potentially unusable 720p h264 encoding to something where you can actually encode 1080p on a 2W cpu with less than 70% cpu usage. Again, In my opinion it is a very big deal.

> I suppose that's true, I have a bad tendency to sort of disregard laptops when it comes to serious streaming -- it's kind of a bad habit.

If I can ask, maybe in the future at least think about disregarding ALL low power cpus and older hardware, not just laptops, since the reality is you can reliably run OBS on this: 



...On Windows.

**Comment by dodgepong** [comment]
There is now a PR that uses the FFmpeg implementation of VAAPI here: https://github.com/obsproject/obs-studio/pull/1255

**Comment by jp9000** [comment]
Moved -> https://github.com/obsproject/obs-studio/pull/1256

**Pull request closed without merging**